### PR TITLE
fix connection uri parsing when the host includes a scheme

### LINKED ISSUE #21638: Spark Connection with k8s in URL not mapped correctly
### Official Helm Chart version

1.2.0

### Apache Airflow version

2.1.4

### Kubernetes Version

v1.21.6+bb8d50a

### Helm Chart configuration

I defined a new Connection String for AIRFLOW_CONN_SPARK_DEFAULT in values.yaml like the following section (base64 encoded or with correct string (spark://k8s://100.68.0.1:443?deploy-mode=cluster):

```
extraSecrets:
  '{{ .Release.Name }}-airflow-connections':
     data: |
       AIRFLOW_CONN_SPARK_DEFAULT: 'c3Bhcms6Ly9rOHM6Ly8xMDAuNjguMC4xOjQ0Mz9kZXBsb3ktbW9kZT1jbHVzdGVy'
```

in Section extraEnvFrom i defined the following:

```
extraEnvFrom: |
   - secretRef:
       name: '{{ .Release.Name }}-airflow-connections'
```

### Docker Image customisations

added apache-airflow-providers-apache-spark to base Image

### What happened

Airflow Connection mapped wrong because of the k8s:// within the url.  if i ask for the connection with cmd "airflow connections get spark_default" then host=k8s and schema=/100.60.0.1:443 which is wrong

### What you expected to happen

the spark Connection based on k8s (spark://k8s://100.68.0.1:443?deploy-mode=cluster) should be parsed correctly

### How to reproduce

define in values.yaml

```
extraSecrets:
  '{{ .Release.Name }}-airflow-connections':
     data: |
       AIRFLOW_CONN_SPARK_DEFAULT: 'c3Bhcms6Ly9rOHM6Ly8xMDAuNjguMC4xOjQ0Mz9kZXBsb3ktbW9kZT1jbHVzdGVy'

extraEnvFrom: |
   - secretRef:
       name: '{{ .Release.Name }}-airflow-connections'
```

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### LINKED ISSUE #31460: Add capability in Airflow connections to validate host
### Apache Airflow version

2.6.1

### What happened

While creating connections in airflow, the form doesn't check for the correctness of the format of the host provided. For instance, we can proceed providing something like this which is not a valid url: `spark://k8s://100.68.0.1:443?deploy-mode=cluster`. It wont instantly fail but will return faulty hosts and other details if called later.

Motivation:
https://github.com/apache/airflow/pull/31376#discussion_r1200112106

### What you think should happen instead

The Connection form can have a validator which should check for these scenarios and report if there is an issue to save developers and users time later.

### How to reproduce

1. Go to airflow connections form
2. Fill in connection host as: `spark://k8s://100.68.0.1:443?deploy-mode=cluster`, other details can be anything
3. Create the connection
4. Run `airflow connections get <name>` 
5. The host and schema will be wrong

### Operating System

Macos

### Versions of Apache Airflow Providers

_No response_

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by hussein-awala**
closes: #21638
closes: #31460

This PR introduces a solution to address the parsing problem that occurs when the connection host includes a scheme.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by uranusjr** [comment]
Instead of trying to guess what the user wants, I’m more inclined to make this fail directly when the user tries to add the connection object in the first place.

**Comment by hussein-awala** [comment]
@uranusjr IMO there should not be any problem if the user provide this kind of hosts, since it's a valid host but with a protocol, for ex, if we create the same connection from the UI:
image

Then we use the Airflow CLI to get the connection information:
```
root@55c0591cd327:/opt/airflow# airflow connections get test_conn
id | conn_id   | conn_type | description | host            | schema | login | password | port | is_encrypted | is_extra_encrypted | extra_dejson       | get_uri                                                   
===+===========+===========+=============+=================+========+=======+==========+======+==============+====================+====================+===========================================================
1  | test_conn | generic   |             | protocol://host |        | user  | pass     | 123  | False        | False              | {'param': 'value'} | generic://user:pass@protocol%3A%2F%2Fhost:123/?param=value
```
We see that it's successfully parsed and we can use it without any problem, and if I think that the method `parse_uri` in the connection class should parse the generated uri and create the same connection from it. WDYT?

**Comment by hussein-awala** [comment]
I will add some changes it to make it backward compatible

**Comment by hussein-awala** [comment]
It's ready and without a new column, I just updated the methods `_parse_from_uri` and `get_uri` and added a new pytest file for the connection model which contain the tests for my changes.

File: airflow/models/connection.py [hunk 1]
```diff
@@ -187,10 +187,22 @@ def _normalize_conn_type(conn_type):
...
     def _parse_from_uri(self, uri: str):
+        schemes_count_in_uri = uri.count("://")
+        if schemes_count_in_uri > 2:
+            raise AirflowException(f"Invalid connection string: {uri}.")
+        scheme_in_uri = schemes_count_in_uri == 2
         uri_parts = urlsplit(uri)
         conn_type = uri_parts.scheme
         self.conn_type = self._normalize_conn_type(conn_type)
-        self.host = _parse_netloc_to_hostname(uri_parts)
+        reset_of_the_url = uri.replace(f"{conn_type}://", ("" if scheme_in_uri else "//"))
```
Comment (amoghrajesh) [inline airflow/models/connection.py]: nit: typo: `rest_of_the_url`

File: airflow/models/connection.py [hunk 2]
```diff
@@ -187,10 +187,22 @@ def _normalize_conn_type(conn_type):
         return conn_type
 
     def _parse_from_uri(self, uri: str):
+        schemes_count_in_uri = uri.count("://")
+        if schemes_count_in_uri > 2:
+            raise AirflowException(f"Invalid connection string: {uri}.")
+        scheme_in_uri = schemes_count_in_uri == 2
```
Comment (amoghrajesh) [inline airflow/models/connection.py]: This variable name is rather confusing to me, can we have something else here?

File: tests/models/test_connection.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+                "spark://k8s://100.68.0.1:443?deploy-mode=cluster",
+                "spark",
+                "k8s://100.68.0.1",
+                None,
+                None,
+                443,
+                "",
+                {"deploy-mode": "cluster"},
+                None,
+            ),
```
Comment (amoghrajesh) [inline tests/models/test_connection.py:98]: Good coverage on the test cases here :)

**Review by amoghrajesh — COMMENTED** [review]
@hussein-awala the changes mostly look good to me.
Have a few comments

**Comment by ephraimbuddy** [comment]
Hi @hussein-awala , we are about to release 2.6.2, can you address the review comments here so this can be part of 2.6.2

File: airflow/models/connection.py (hunk 2 above)
Comment (hussein-awala) [reply airflow/models/connection.py]: can you check it now?

**Review by amoghrajesh — APPROVED** [review]
Thanks for those changes, @hussein-awala
LGTM +1

**Comment by hussein-awala** [comment]
> Hi @hussein-awala , we are about to release 2.6.2, can you address the review comments here so this can be part of 2.6.2

@ephraimbuddy the review is addressed, can your review the PR?

**Review by ephraimbuddy — APPROVED** [review]

**Comment by ephraimbuddy** [comment]
Hi @uranusjr , do you want to take another look?

**Review by uranusjr — APPROVED** [review]

**Pull request merged**